src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,377 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+#nullable enable
+
+namespace System.Reflection.Metadata
+{
+    internal static class TypeNameParserHelpers
+    {
+        internal const sbyte SZArray = -1;
+        internal const sbyte Pointer = -2;
+        internal const sbyte ByRef = -3;
+        private const char EscapeCharacter = '\\';
+#if NET8_0_OR_GREATER
+        private static readonly SearchValues<char> s_endOfFullTypeNameDelimitersSearchValues = SearchValues.Create("[]&*,+\\");
+#endif
+
+        internal static string GetGenericTypeFullName(ReadOnlySpan<char> fullTypeName, ReadOnlySpan<TypeName> genericArgs)
+        {
+            Debug.Assert(genericArgs.Length > 0);
+
+            ValueStringBuilder result = new(stackalloc char[128]);
```

Comment:
This can be made more efficient by creating one top level ValueStringBuilder instead of creating ValueStringBuilder at each level of the generic type recursion.

Reply:
You are right. But after I've addressed https://github.com/dotnet/runtime/pull/100094#discussion_r1577317365 `FullName` is very unlikely to be used for generic types. Since it should be rare I am not going to implement it for now.